### PR TITLE
Avoid to miss to root for local slots before the hard fork

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

some rooted slots could be omitted to be marked.

some rooted slots could be omitted after restart (hard fork), if the hard fork slot is ahead of the validator's latest ledger root before the restart.

This is usually the case because current practice of cluster restart is based on the opt-conf-ed slot. (i.e. no node should have rooted that slot yet)

ultimately, this leads to hole in bigtable.

cc: @mvines @sakridge @carllin 

#### Summary of Changes

~~fix it. (this is wip; I'm working on to fix the actual hole now).~~ impl finished

related: #13536

### CODE REVIEW

File: core/src/validator.rs [hunk 1]
```diff
@@ -1266,6 +1289,24 @@ fn new_banks_from_ledger(
         );
     }
 
+    if let Some(hard_fork_restart_slot) =
+        maybe_cluster_restart_with_hard_fork(config, bank_forks.root_bank().slot())
+    {
+        if !config.no_hard_fork_blockstore_root_reconcilation_for_local_cluster_test {
```
Comment (ryoqun) [inline core/src/validator.rs:1295]: I have to swallow this impurity for the completeness of integration tests (and lack of unit tests around these code). xD

File: core/src/validator.rs [hunk 2]
```diff
@@ -1266,6 +1289,24 @@ fn new_banks_from_ledger(
...
+            reconcile_blockstore_roots_with_external_source(
+                ExternalRootSource::HardFork(hard_fork_restart_slot),
+                &blockstore,
+                &mut last_blockstore_root,
+            )
```
Comment (ryoqun) [inline core/src/validator.rs:1300]: so this is the meat of this pr

File: local-cluster/src/local_cluster.rs [hunk 3]
```diff
@@ -128,6 +134,23 @@ impl LocalCluster {
         Self::new(&mut config, socket_addr_space)
     }
 
+    fn sync_ledger_path_across_nested_config_fields(
+        config: &mut ValidatorConfig,
+        ledger_path: &Path,
+    ) {
+        config.account_paths = vec![ledger_path.join("accounts")];
+        config.tower_storage = Arc::new(FileTowerStorage::new(ledger_path.to_path_buf()));
```
Comment (ryoqun) [inline local-cluster/src/local_cluster.rs:142]: there were three instances of this code. it took some time to debug...

File: local-cluster/tests/local_cluster.rs [hunk 4]
```diff
@@ -3762,6 +3786,222 @@ fn test_run_test_load_program_accounts_partition_root() {
...
+    let validators = validator_keys
+        .iter()
+        .map(|(kp, _)| kp.pubkey())
+        .collect::<Vec<_>>();
+
+    let validator_a_pubkey = validators[0];
+    let validator_b_pubkey = validators[1];
+
+    let validator_config = ValidatorConfig {
+        snapshot_config: Some(LocalCluster::create_dummy_load_only_snapshot_config()),
```
Comment (ryoqun) [inline local-cluster/tests/local_cluster.rs:3867]: this was crucial actually to load from a snapshot....

File: core/src/consensus.rs [hunk 5]
```diff
@@ -1221,45 +1221,63 @@ impl TowerError {
...
             })
             .collect();
         if !new_roots.is_empty() {
             info!(
-                "Reconciling slots as root based on tower root: {:?} ({}..{}) ",
-                new_roots, tower_root, last_blockstore_root
+                "Reconciling slots as root based on external root: {:?} ({}..{}) ",
+                new_roots, external_root, last_blockstore_root
             );
             blockstore.set_roots(new_roots.iter())?;
```
Comment (ryoqun) [inline core/src/consensus.rs:1270]: @carllin I wonder this `.set_roots(...)` should be accompanied with `set_duplicate_confirmed_slots_and_hashes(...)` like in blockstore_processor?: https://github.com/solana-labs/solana/blob/5bbb0da7b80448655059465704beb9fee5183ab8/ledger/src/blockstore_processor.rs#L1169

Comment (carllin) [reply core/src/consensus.rs:1270]: @ryoqun ah yeah it should, good catch!

Although I think you'll see that there are no calls to `fn set_duplicate_confirmed_slots_and_hashes` in 1.6 and 1.7 currently 😄

File: core/src/consensus.rs [hunk 6]
```diff
@@ -1221,45 +1221,63 @@ impl TowerError {
...
+    external_source: ExternalRootSource,
     blockstore: &Blockstore,
+    last_blockstore_root: &mut Slot,
 ) -> blockstore_db::Result<()> {
-    let tower_root = tower.root();
-    let last_blockstore_root = blockstore.last_root();
-    if last_blockstore_root < tower_root {
-        // Ensure tower_root itself to exist and be marked as rooted in the blockstore
+    let external_root = external_source.root();
+    if *last_blockstore_root < external_root {
```
Comment (carllin) [inline core/src/consensus.rs:1250]: ah subtle, so if the hard fork is at a slot greater than this validator's last root, i.e. this validator was behind, then we can miss some roots. Great catch!

Want to add this to the PR problem description, i.e. "some rooted slots could be omitted after restart (hard fork), if the hard fork slot is ahead of the validator's latest ledger root before the restart"

File: core/src/consensus.rs [hunk 7]
```diff
@@ -1221,45 +1221,63 @@ impl TowerError {
...
             .collect();
         if !new_roots.is_empty() {
             info!(
-                "Reconciling slots as root based on tower root: {:?} ({}..{}) ",
-                new_roots, tower_root, last_blockstore_root
+                "Reconciling slots as root based on external root: {:?} ({}..{}) ",
+                new_roots, external_root, last_blockstore_root
             );
             blockstore.set_roots(new_roots.iter())?;
+            *last_blockstore_root = blockstore.last_root();
```
Comment (carllin) [inline core/src/consensus.rs:1271]: Is this change necessary?

File: local-cluster/tests/local_cluster.rs [hunk 8]
```diff
@@ -3762,6 +3786,222 @@ fn test_run_test_load_program_accounts_partition_root() {
...
+        );
+        assert_eq!(
+            (slots_a.last().unwrap(), roots_a.last().unwrap()),
+            (slots_b.last().unwrap(), roots_b.last().unwrap())
+        );
+        assert_ne!((slots_a, roots_a), (slots_b, roots_b));
+    }
+}
+
+// the following two tests are rather fragile, depending on successful loading of the hard-forked snapshot.
```
Comment (carllin) [inline local-cluster/tests/local_cluster.rs:3990]: Could you add some more comments expanding on what scenarios the tests are fragile under/expected failures?

File: core/src/validator.rs [hunk 9]
```diff
@@ -158,6 +158,7 @@ pub struct ValidatorConfig {
     pub validator_exit: Arc<RwLock<Exit>>,
     pub no_wait_for_vote_to_start_leader: bool,
     pub accounts_shrink_ratio: AccountShrinkThreshold,
+    pub no_hard_fork_blockstore_root_reconcilation_for_local_cluster_test: bool,
```
Comment (carllin) [inline core/src/validator.rs:161]: Maybe a comment on why this is necessary?

File: local-cluster/tests/local_cluster.rs [hunk 10]
```diff
@@ -3762,6 +3786,222 @@ fn test_run_test_load_program_accounts_partition_root() {
...
+#[serial]
+fn test_hard_fork_without_gap_in_roots() {
+    do_test_hard_fork_with_or_without_gap_in_roots(false);
+}
+
+#[test]
+#[serial]
+fn test_hard_fork_with_gap_in_roots() {
+    do_test_hard_fork_with_or_without_gap_in_roots(true);
+}
```
Comment (carllin) [inline local-cluster/tests/local_cluster.rs:4003]: I think it might be ok to just have the success test `test_hard_fork_without_gap_in_roots`, without the need to test expecting an explicit failure that we've already fixed. I think this also removes the need for the additional `ValidatorConfig` flag

File: local-cluster/tests/local_cluster.rs [hunk 11]
```diff
@@ -3762,6 +3786,222 @@ fn test_run_test_load_program_accounts_partition_root() {
...
+            {
+                break;
+            }
+        }
+    }
+
+    let mut validator_a_info = cluster.lock().unwrap().exit_node(&validator_a_pubkey);
+    let mut validator_b_info = cluster.lock().unwrap().exit_node(&validator_b_pubkey);
+
+    let hard_fork_slot = min_last_vote - 5;
```
Comment (carllin) [inline local-cluster/tests/local_cluster.rs:3906]: maybe an assertion here after node exit that the `hard_fork_slot` is greater than the root_in_tower, to ensure no weird race condition happened after node exit

**Comment by ryoqun** [comment]
it looks like this is working this time

```
[2021-10-06T08:12:51.669593754Z INFO  solana_ledger::blockstore_processor] ledger processed in 358 ms, 698 µs and 305 ns. root slot is 96965687, 1 fork at 96965687, with 1 frozen bank
[2021-10-06T08:12:51.670081648Z INFO  solana_core::consensus] Reconciling slots as root based on external root: [96965687, 96965686, 96965685, 96965684, 96965655, 96965654, 96965653, 96965652, 96965643, 96965642, 96965641, 96965635, 96965634, 96965633, 96965632, 96965595, 96965594, 96965593, 96965592, 96965583, 96965582, 96965580, 96965579, 96965578, 96965577, 96965576, 96965575, 96965574, 96965573, 96965572, 96965563, 96965562, 96965561, 96965560, 96965559, 96965558, 96965557, 96965556, 96965547, 96965546, 96965543, 96965542, 96965541, 96965537, 96965536, 96965535, 96965534, 96965533, 96965532, 96965531, 96965530, 96965529, 96965528, 96965523, 96965522, 96965521, 96965520, 96965511, 96965510, 96965509, 96965508, 96965495, 96965494, 96965492, 96965491, 96965490, 96965489, 96965488, 96965483, 96965482, 96965481, 96965480, 96965455, 96965454, 96965453, 96965452, 96965451, 96965450, 96965449, 96965448, 96965439, 96965438, 96965437, 96965436, 96965435, 96965434, 96965433, 96965432, 96965423, 96965422, 96965421, 96965420, 96965415, 96965414, 96965413, 96965412, 96965411, 96965410, 96965409, 96965408] (96965687..96965407) 
[2021-10-06T08:12:51.670466083Z INFO  solana_core::consensus] adjusting lockouts (after replay up to 96965687): [96965411, 96965412, 96965432, 96965433, 96965434, 96965435, 96965436, 96965448, 96965449, 96965450, 96965480, 96965481, 96965488, 96965489, 96965490, 96965508, 96965520, 96965521, 96965528, 96965529, 96965530, 96965531, 96965533, 96965541, 96965716, 96965717, 96965718, 96965719] tower root: 96965407 replayed root: 96965687

```

**Comment by ryoqun** [comment]
(time to revive this pr finally......)

**Comment by CriesofCarrots** [comment]
@ryoqun , any chance to resuscitating this??

**Comment by ryoqun** [comment]
status: I resuscitated this pr with all review so far being addressed. I'll write replies to review comments later.

File: core/src/consensus.rs [hunk 12]
```diff
@@ -1320,45 +1320,77 @@ impl TowerError {
...
+
+            // Unfortunately, we can't supply duplicate-confirmed hashes,
+            // because it can't be guaranteed to be able to replay these slots
+            // under this code-path's limited condition (i.e.  those shreds
+            // might not be available, etc...) also correctly overcoming this
+            // limitation is hard...
+            blockstore.mark_slots_as_if_rooted_normally_at_startup(
+                new_roots.into_iter().map(|root| (root, None)).collect(),
+                false,
+            )?;
```
Comment (ryoqun) [inline core/src/consensus.rs:1379]: here

File: core/src/consensus.rs (hunk 5 above)
Comment (ryoqun) [reply core/src/consensus.rs]: after some code reading, i had to make a kind of compromise regarding the treatment of `duplicate confirmed slots`...: https://github.com/solana-labs/solana/pull/19912/files#r902224585

so, i've defined small `fn` (`mark_slots_as_if_rooted_normally_at_startup`) for the sole sake of documentation to make the nuance very explicit, to my taste. i hope it'll be more immune to comment rots. ;)

File: core/src/consensus.rs (hunk 6 above)
Comment (ryoqun) [reply core/src/consensus.rs:1351]: yeah, i updated pr problem

File: core/src/consensus.rs [hunk 13]
```diff
@@ -1320,45 +1320,77 @@ impl TowerError {
...
 // between tower (plain old POSIX fs calls) and blockstore (through RocksDB), when
 // `ReplayState::handle_votable_bank()` saves tower before setting blockstore roots.
-pub fn reconcile_blockstore_roots_with_tower(
-    tower: &Tower,
+pub fn reconcile_blockstore_roots_with_external_source(
+    external_source: ExternalRootSource,
     blockstore: &Blockstore,
+    // blockstore.last_root() might have been updated already.
+    // so take a &mut param both to input (and output iff we update root)
+    last_blockstore_root: &mut Slot,
```
Comment (ryoqun) [inline core/src/consensus.rs:1348]: there

File: core/src/consensus.rs (hunk 7 above)
Comment (ryoqun) [reply core/src/consensus.rs:1384]: yeah, sort of. i add comment here and [there](https://github.com/solana-labs/solana/pull/19912/files#r902227844).

File: core/src/consensus.rs [hunk 14]
```diff
@@ -1221,45 +1221,63 @@ impl TowerError {
...
+            .take_while(|current| match current.cmp(last_blockstore_root) {
                 Ordering::Greater => true,
                 Ordering::Equal => false,
                 Ordering::Less => panic!(
-                    "couldn't find a last_blockstore_root upwards from: {}!?",
-                    tower_root
+                    "last_blockstore_root({}) is skipped while traversing blockstore (currently at {}) from external root ({})!?",
+                    last_blockstore_root,
+                    current,
+                    external_root
```
[the remark this answers is not in the document]
Comment (ryoqun) [reply core/src/consensus.rs]: done.

File: local-cluster/tests/local_cluster.rs (hunk 8 above)
Comment (ryoqun) [reply local-cluster/tests/local_cluster.rs]: this got n/a because i've simplified the tests.

File: core/src/validator.rs (hunk 9 above)
Comment (ryoqun) [reply core/src/validator.rs]: this got n/a because i've simplified the tests.

File: local-cluster/tests/local_cluster.rs (hunk 10 above)
Comment (ryoqun) [reply local-cluster/tests/local_cluster.rs]: yeah, i made so.

File: local-cluster/tests/local_cluster.rs (hunk 11 above)
Comment (ryoqun) [reply local-cluster/tests/local_cluster.rs:2271]: done.

File: core/src/tvu.rs [hunk 15]
```diff
@@ -430,7 +429,7 @@ pub mod tests {
                 OptimisticallyConfirmedBank::locked_from_bank_forks_root(&bank_forks),
             )),
             &poh_recorder,
-            tower,
+            None,
```
Comment (ryoqun) [inline core/src/tvu.rs:432]: here

File: core/src/replay_stage.rs [hunk 16]
```diff
@@ -375,8 +376,14 @@ impl ReplayStage {
...
         transaction_cost_metrics_sender: Option<TransactionCostMetricsSender>,
     ) -> Self {
-        let mut tower = tower.into();
-        info!("Tower state: {:?}", tower);
+        let mut tower = if let Some(process_blockstore) = maybe_process_blockstore {
+            let tower = process_blockstore.process_to_create_tower();
+            info!("Tower state: {:?}", tower);
+            tower
+        } else {
+            warn!("creating default tower....");
```
Comment (ryoqun) [inline core/src/replay_stage.rs:384]: here

File: core/src/validator.rs [hunk 17]
```diff
@@ -1639,15 +1663,30 @@ impl<'a> ProcessBlockStore<'a> {
...
+                    abort()
+                });
+            }
+
             *self.start_progress.write().unwrap() = previous_start_process;
         }
     }
-}
 
-impl<'a> From<ProcessBlockStore<'a>> for Tower {
```
Comment (ryoqun) [inline core/src/validator.rs:1647]: @mvines I think this isn't correct use of `From`, which is introduced at #23852 (https://github.com/solana-labs/solana/pull/23852/files#r902237878).

This is converting quite different things. ideally, `From` should be a cast for conceptually similar things. Also, this made code reading a bit hard because i can't easily spot when `ProcessBlockStore::process` is exactly executed now. (i.e. searching the correct `.into()` is hard)

it seems this trait workaround is only needed for this test [test_tvu_exit](https://github.com/solana-labs/solana/pull/19912/files#r902242642) and i think alternative approach should arguably be tolerable (still not ideal, i skipped proper introducing enum or trait just for this purpose...): https://github.com/solana-labs/solana/pull/19912/files#r902243960

btw, i didn't know there is an identity/no-op blanket `From`/`Into` impl provided here
https://doc.rust-lang.org/src/core/convert/mod.rs.html#559:

```rust
// From (and thus Into) is reflexive
#[stable(feature = "rust1", since = "1.0.0")]
#[rustc_const_unstable(feature = "const_convert", issue = "88674")]
impl<T> const From<T> for T {
    /// Returns the argument unchanged.
    fn from(t: T) -> T {
        t
    }
}
```


granted, this refactor isn't relevant to this pr but i couldn't resist the urge, when i was code-reading to guarantee the safety of the scary blockstore mutation with hard fork params.

happy to separate this change into another preparatory pr, if you desire. :)

**Comment by ryoqun** [comment]
> status: ... I'll write replies to review comments later.

this is finally done!

@carllin could you review this? cc: @CriesofCarrots

File: core/src/validator.rs (hunk 17 above)
Comment (mvines) [reply core/src/validator.rs:1647]: Sure, this looks fine to me

File: core/src/consensus.rs (hunk 12 above)
Comment (carllin) [reply core/src/consensus.rs:1379]: hmmm yeah this is yet another tricky edge case to reason about, but I think it should be ok.

 Not marking these slots as duplicate confirmed just means:

1. we can't serve requests to people asking for the correct version of these slots via `AncestorHashesRepairType`
2. Should be safe as long as we're not freezing the slot again during replay (not happening b/c rooted)

**Review by carllin — APPROVED** [review]

**Comment by ryoqun** [comment]
sanity checked and got:

```
$ ./target/release/solana-ledger-tool -l config/ledger bounds
[2022-06-26T05:34:56.257536980Z INFO  solana_ledger_tool] solana-ledger-tool 1.11.0 (src:34ad3d51; feat:528226107)
[2022-06-26T05:34:56.257639239Z INFO  solana_ledger::blockstore] Maximum open file descriptors: 500000
[2022-06-26T05:34:56.257647743Z INFO  solana_ledger::blockstore] Opening database at "/home/ryoqun/work/solana/solana/config/ledger/rocksdb"
[2022-06-26T05:34:56.257664842Z INFO  solana_ledger::blockstore_db] Disabling rocksdb's automatic compactions...
[2022-06-26T05:34:56.258457406Z INFO  solana_ledger::blockstore_db] Opening Rocks with secondary (read only) access at: "/home/ryoqun/work/solana/solana/config/ledger/rocksdb/solana-secondary"
[2022-06-26T05:34:56.258475519Z INFO  solana_ledger::blockstore_db] This secondary access could temporarily degrade other accesses, such as by solana-validator
[2022-06-26T05:34:56.293056701Z INFO  solana_ledger::blockstore] "/home/ryoqun/work/solana/solana/config/ledger/rocksdb" open took 35ms
Ledger has data for 225 slots 0 to 224
  with 158 rooted slots from 0 to 192
  and 32 slots past the last root


...

[2022-06-26T05:35:39.983289202Z INFO  solana_ledger::blockstore_processor] ledger processed in 157 µs and 890 ns. root slot is 223, 1 bank: 223
[2022-06-26T05:35:39.983289475Z INFO  solana_metrics::metrics] datapoint: process_blockstore_from_root total_time_us=157i frozen_banks=1i slot=223i forks=1i calculate_capitalization_us=23321i
[2022-06-26T05:35:39.983590734Z INFO  solana_core::consensus] adjusting lockouts (after replay up to 223): [193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223] tower root: 192 replayed root: 223
[2022-06-26T05:35:39.983611982Z INFO  solana_core::consensus] adjusted tower's anchored slot: Some(223)
[2022-06-26T05:35:39.983616519Z INFO  solana_core::consensus] All restored votes were behind; resetting root_slot and last_vote in tower!
[2022-06-26T05:35:39.983665577Z ERROR solana_core::validator] Hard fork is detected; discarding tower restoration result: Ok(Tower { node_pubkey: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm, threshold_depth: 8, threshold_size: 0.6666666666666666, vote_state: VoteState { node_pubkey: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7h
gBhAN68zVm, authorized_withdrawer: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm, commission: 44, votes: [], root_slot: Some(223), authorized_voters: AuthorizedVoters { authorized_voters: {0: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm} }, prior_voters: CircBuf { buf: [(11111111111111111111111111111111, 0, 0), (111111
11111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (1111111111111111111111111111111
1, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111
111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0)
, (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (1111111111111111111111
1111111111, 0, 0)], idx: 31, is_empty: true }, epoch_credits: [(0, 156, 0)], last_timestamp: BlockTimestamp { slot: 0, timestamp: 0 } }, last_vote: Vote(Vote { slots: [], hash: 11111111111111111111111111111111, timestamp: None }), last_vote_tx_blockhash: 11111111111111111111111111111111, last_timestamp: BlockTimestamp
 { slot: 222, timestamp: 1656221556 }, stray_restored_slot: None, last_switch_threshold_check: None })
[2022-06-26T05:35:39.983707287Z ERROR solana_core::validator] Rebuilding a new tower from the latest vote account due to failed tower restore: The tower is useless because of new hard fork: 223
[2022-06-26T05:35:39.983671541Z ERROR solana_metrics::metrics] datapoint: tower_error error="Hard fork is detected; discarding tower restoration result: Ok(Tower { node_pubkey: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm, threshold_depth: 8, threshold_size: 0.6666666666666666, vote_state: VoteState { node_pubkey: 7eV
T4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm, authorized_withdrawer: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm, commission: 44, votes: [], root_slot: Some(223), authorized_voters: AuthorizedVoters { authorized_voters: {0: 7eVT4kj1gJ7dij8bDEeqmAYcWPP7jyDV7hgBhAN68zVm} }, prior_voters: CircBuf { buf: [(11111111111111111
111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (
11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (1111111111111111111111111
1111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111
111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111, 0, 0), (11111111111111111111111111111111,
 0, 0), (11111111111111111111111111111111, 0, 0)], idx: 31, is_empty: true }, epoch_credits: [(0, 156, 0)], last_timestamp: BlockTimestamp { slot: 0, timestamp: 0 } }, last_vote: Vote(Vote { slots: [], hash: 11111111111111111111111111111111, timestamp: None }), last_vote_tx_blockhash: 11111111111111111111111111111111,
 last_timestamp: BlockTimestamp { slot: 222, timestamp: 1656221556 }, stray_restored_slot: None, last_switch_threshold_check: None })"
[2022-06-26T05:35:39.983734623Z ERROR solana_metrics::metrics] datapoint: tower_error error="Unable to restore tower: The tower is useless because of new hard fork: 223"
[2022-06-26T05:35:39.983949656Z INFO  solana_core::consensus] Reconciling slots as root based on external root: [223, 222, 221, 220, 219, 218, 217, 216, 215, 214, 213, 212, 211, 210, 209, 208, 207, 206, 205, 204, 203, 202, 201, 200, 199, 198, 197, 196, 195, 194, 193] (external: HardFork(223), blockstore: 192)
[2022-06-26T05:35:39.984036580Z INFO  solana_core::validator] Waiting for 80% of activated stake at slot 223 to be in gossip...
[2022-06-26T05:35:39.984046941Z INFO  solana_core::validator] Supermajority reached, 100% active stake detected, starting up now.
```

seems working properly.

**Comment by ryoqun** [comment]
seems this is worked for 2023-02-26 mainnet-beta outage:

```
$ grep -E "Reconcil|storage-bigtable-upload-block.*17952640" logs/solana-validator.log | sort -k 6
[2023-02-26T01:36:36.707933988Z INFO  solana_core::consensus] Reconciling slots as root based on external root: [179526403, 179526402, 179526401, 179526400] (external: HardFork(179526403), blockstore: 179526391)
[2023-02-26T01:59:17.685949819Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526400i transactions=43i bytes=84527i
[2023-02-26T01:59:22.383623281Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526401i transactions=3353i bytes=1601930i
[2023-02-26T01:59:19.889260232Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526402i transactions=2400i bytes=1395552i
[2023-02-26T01:59:19.276003004Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526403i transactions=2335i bytes=857771i
[2023-02-26T01:59:17.685875247Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526404i transactions=0i bytes=116i
[2023-02-26T01:59:17.686201328Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526405i transactions=958i bytes=391503i
[2023-02-26T01:59:17.686394797Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526406i transactions=1460i bytes=574413i
[2023-02-26T01:59:17.685981530Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526407i transactions=254i bytes=139982i
[2023-02-26T01:59:17.686645294Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526408i transactions=2715i bytes=1006228i
[2023-02-26T01:59:17.686236445Z INFO  solana_metrics::metrics] datapoint: storage-bigtable-upload-block slot=179526409i transactions=906i bytes=370063i

```

File: local-cluster/tests/local_cluster.rs [hunk 18]
```diff
@@ -2120,6 +2146,224 @@ fn test_run_test_load_program_accounts_root() {
...
+    let slots_b = AncestorIterator::new(common_last_vote, &blockstore_b).collect::<Vec<_>>();
+    let roots_b = blockstore_b
+        .reversed_rooted_slot_iterator(common_root)
+        .unwrap()
+        .collect::<Vec<_>>();
+
+    // compare them all!
+    assert_eq!((&slots_a, &roots_a), (&slots_b, &roots_b));
+    assert_eq!(&slots_a[slots_a.len() - roots_a.len()..].to_vec(), &roots_a);
+    assert_eq!(&slots_b[slots_b.len() - roots_b.len()..].to_vec(), &roots_b);
```
Comment (ryoqun) [inline local-cluster/tests/local_cluster.rs:2364]: here

File: ledger/src/blockstore_meta.rs [hunk 19]
```diff
@@ -214,6 +214,11 @@ impl SlotMeta {
         Some(self.consumed) == self.last_index.map(|ix| ix + 1)
     }
 
+    /// Dangerous. Currently only needed for a local-cluster test
+    pub fn unset_parent(&mut self) {
+        self.parent_slot = None;
+    }
+
```
Comment (ryoqun) [inline ledger/src/blockstore_meta.rs:221]: https://github.com/solana-labs/solana/pull/32427
